featuregate/flag.go

```diff
@@ -5,40 +5,88 @@ package featuregate // import "go.opentelemetry.io/collector/featuregate"
 
 import (
 	"flag"
+	"fmt"
 	"strings"
 
 	"go.uber.org/multierr"
 )
 
 // NewFlag returns a flag.Value that directly applies feature gate statuses to a Registry.
+//
+// Deprecated: use NewFlags instead to get both gates and strict flags.
 func NewFlag(reg *Registry) flag.Value {
-	return &flagValue{reg: reg}
+	return &gatesFlagValue{args: &gateRegistrationArgs{reg: reg}}
 }
 
-// flagValue implements the flag.Value interface and directly applies feature gate statuses to a Registry.
-type flagValue struct {
-	reg *Registry
+// NewFlags returns two flag.Values: one that registers gates, and one that sets strict mode.
+func NewFlags(reg *Registry) (flag.Value, flag.Value) {
+	g := &gateRegistrationArgs{reg: reg}
+	return &gatesFlagValue{args: g}, &strictFlagValue{args: g}
 }
 
-func (f *flagValue) String() string {
-	var ids []string
-	f.reg.VisitAll(func(g *Gate) {
-		id := g.ID()
-		if !g.IsEnabled() {
+// gatesFlagValue implements the flag.Value interface and records gates to set on the Registry.
+type gatesFlagValue struct {
+	args *gateRegistrationArgs
+}
+
+func (f *gatesFlagValue) String() string {
+	var gates []string
+	f.args.reg.VisitAll(func(gate *Gate) {
+		id := gate.ID()
+		if !gate.IsEnabled() {
 			id = "-" + id
 		}
-		ids = append(ids, id)
+		gates = append(gates, id)
 	})
-	return strings.Join(ids, ",")
+	return strings.Join(gates, ",")
 }
 
-func (f *flagValue) Set(s string) error {
-	if s == "" {
-		return nil
+func (f *gatesFlagValue) Set(s string) error {
+	f.args.gates = s
```

Comment:
Does this prevent the `--feature-gates` flag from being provided multiple times?

Reply:
No, though that was never supported from what I can see.

Reply:
Can you add a test that proves this?  Implementing the `flag.Value` interface should allow the flag to be specified multiple times on the CLI at invocation, resulting in a call to `Set()` for each instance of the flag.

Reply:
https://go.dev/play/p/wQG2ktm4uQE demonstrates the concept of having a `flag.Value` specified multiple times.